app/src/main/java/eu/faircode/email/HtmlHelper.java

```diff
@@ -60,7 +60,7 @@
 
 public class HtmlHelper {
     private static final int PREVIEW_SIZE = 250;
-    private static Pattern pattern = Pattern.compile("([http|https]+://[\\w\\S(\\.|:|/)]+)");
+    private static final Pattern pattern = Pattern.compile("https?://[\\w#%&()*+,\\-./:;=?@\\[\\]_|~]*[\\w#)*+\\-/=@\\]|~]");
```

Comment:
You'll have to explain at least how this regex works / is different from the original regex and why. If you got this regex from somewhere, you'll have to tell where it comes from.

Regex'es like this are notoriously hard to get right, which is why I have kept it as simple as possible. It is just an attempt to auto link 95% of the URLs without failing hard or running into all kind of edge cases.

Reply:
The new regular expression is partly explained by the newly added unit tests. If you change the regular expression back to its previous version, it will fail some of the tests.

The regular expression consists of 3 parts:
1. `https?://`
2. `[\w#%&()*+,\-./:;=?@\[\]_|~]*`
3. `[\w#)*+\-/=@\]|~]`

Part 1 should be obvious. It's either `http` or `https`.

Part 2 matches any ASCII character that is either allowed in URLs or is commonly found in URLs. That's digits and letters, plus several punctuation characters, ordered by their Unicode code point. Some example characters that are not included here are `<>` and `'"` since these are usually used to separate a URL from the surrounding text. The characters `()` are included though, because when copying a Wikipedia URL, the trailing `)` is not percent-encoded. This last case is debatable. The `)` should probably only be part of the URL if there already is a corresponding `(` in the URL. But then it gets complicated …

Part 3 is a subset of part 2. It is missing characters which usually don't appear at the very end of a URL but are more likely punctuation from the surrounding text. Examples for these are dot, comma, colon.

Reply:
Matching URIs in a text is problematic and will always have edge cases, so I think I will just switch to Patterns.WEB_URL because it will be *very* hard to do any better.

Source code:
https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/util/Patterns.java

The text replacing can be rewritten by creating Jsoup node elements, which will solve any encoding problems.

Reply:
Your choice. I just had a look at the test cases from the [PatternsTest](https://android.googlesource.com/platform/frameworks/base/+/master/core/tests/coretests/src/android/util/PatternsTest.java), and they were missing a few that I have in my list. The `WEB_URL` pattern fails at Typo3 and Wikipedia URLs, and it removes a trailing `/` even though it shouldn't.

Reply:
I am sure there are plenty of other edge cases and that it won't ever be perfect, so choices have to be made. I think Patterns.WEB_URL is as good as it can get.

Reply:
Thanks for trying to improve this anyway, much appreciated!